README.md
ktx
[![crates.io](https://img.shields.io/crates/v/ktx.svg)](https://crates.io/crates/ktx)
[![Documentation](https://docs.rs/ktx/badge.svg)](https://docs.rs/ktx)
==========

KTX v1 texture storage format parsing.

Parses byte data according to [https://www.khronos.org/registry/KTX/specs/1.0/ktxspec_v1.html](https://www.khronos.org/registry/KTX/specs/1.0/ktxspec_v1.html).

```rust
// Include & use static ktx data
use ktx::{Ktx, include_ktx, KtxInfo};

let image: Ktx<_> = include_ktx!("../tests/babg-bc3.ktx");
assert_eq!(image.pixel_width(), 260);
```

```rust
// Read ktx data at runtime
use ktx::KtxInfo;

let decoder = ktx::Decoder::new(buf_reader)?;
assert_eq!(decoder.pixel_width(), 260);
```

## Minimum supported rust compiler
This crate is maintained with [latest stable rust](https://gist.github.com/alexheretic/d1e98d8433b602e57f5d0a9637927e0c).


CHANGELOG.md
# 0.3.2
* Support cubemap textures. For such textures each level will contain all 6 faces in order: +X, -X, +Y, -Y, +Z, -Z.

# 0.3.1
* Add new default feature `std` which can be disabled to allow no-std usage.

# 0.3
* `Ktx` now uses a generic `D: Deref<Target = [u8]>` instead of requiring a slice. This supports owned data usage in addition to slice usage. The texture slice lifetime is now the shorter `Ktx` lifetime because of this.
* Remove internal macros instead `AsRef<KtxHeader>` now provides `KtxInfo` allowing `Ktx`, `KtxHeader`, `KtxDecoder` to all implement `KtxInfo`.

# 0.2
* Add `ktx::Decoder` useful when reading from a file and/or compressed data.
* Separate header logic into `KtxInfo` trait now provided by methods instead of direct field access.

# 0.1
Initial release supporting KTX formatted byte data.

tests/issues.rs
use ktx::*;
use std::{fs::File, io::BufReader};

// #9 panic due to buggy? file texture indices.
#[test]
fn issue9_panic_reading_uffizi_textures() {
    let texture = include_ktx!("uffizi_rgba16f_cube.ktx");

    dbg!(texture);
    assert_eq!(texture.mipmap_levels(), 10);

    assert_eq!(texture.textures().count(), 10);

    for data in texture.textures() {
        dbg!(data.len());
    }
}

// #9 no panic in runtime read version, but worth keeping for regression testing.
#[test]
fn issue9_panic_reading_uffizi_textures_runtime() {
    let ktx_file = BufReader::new(File::open("tests/uffizi_rgba16f_cube.ktx").unwrap());
    let texture = ktx::Decoder::new(ktx_file).unwrap();

    dbg!(&texture);

    assert_eq!(texture.mipmap_levels(), 10);

    let levels: Vec<_> = texture.read_textures().collect();

    assert_eq!(levels.len(), 10);

    for data in levels {
        dbg!(data.len());
    }
}


tests/example.rs
use blake2::{Blake2s, Digest};
use ktx::*;
use std::{
    fs::File,
    io::{self, BufReader},
    sync::Arc,
};

#[test]
fn include_logo_example() {
    let ktx = include_ktx!("babg-bc3.ktx");

    assert!(!ktx.big_endian(), "!big_endian");
    assert_eq!(ktx.gl_type(), 0, "gl_type");
    assert_eq!(ktx.gl_type_size(), 1, "gl_type_size");
    assert_eq!(ktx.gl_format(), 0, "gl_format");
    assert_eq!(ktx.gl_internal_format(), 33779, "gl_internal_format");
    assert_eq!(
        ktx.gl_base_internal_format(),
        6408,
        "gl_base_internal_format"
    );
    assert_eq!(ktx.pixel_width(), 260, "pixel_width");
    assert_eq!(ktx.pixel_height(), 200, "pixel_height");
    assert_eq!(ktx.pixel_depth(), 0, "pixel_depth");
    assert_eq!(ktx.array_elements(), 0, "array_elements");
    assert_eq!(ktx.faces(), 1, "faces");
    assert_eq!(ktx.mipmap_levels(), 8, "mipmap_levels");
    assert_eq!(ktx.bytes_of_key_value_data(), 0, "bytes_of_key_value_data");
}

#[test]
fn read_logo_example() -> io::Result<()> {
    let ktx = ktx::Decoder::new(BufReader::new(File::open("tests/babg-bc3.ktx")?))?;

    assert!(!ktx.big_endian(), "!big_endian");
    assert_eq!(ktx.gl_type(), 0, "gl_type");
    assert_eq!(ktx.gl_type_size(), 1, "gl_type_size");
    assert_eq!(ktx.gl_format(), 0, "gl_format");
    assert_eq!(ktx.gl_internal_format(), 33779, "gl_internal_format");
    assert_eq!(
        ktx.gl_base_internal_format(),
        6408,
        "gl_base_internal_format"
    );
    assert_eq!(ktx.pixel_width(), 260, "pixel_width");
    assert_eq!(ktx.pixel_height(), 200, "pixel_height");
    assert_eq!(ktx.pixel_depth(), 0, "pixel_depth");
    assert_eq!(ktx.array_elements(), 0, "array_elements");
    assert_eq!(ktx.faces(), 1, "faces");
    assert_eq!(ktx.mipmap_levels(), 8, "mipmap_levels");
    assert_eq!(ktx.bytes_of_key_value_data(), 0, "bytes_of_key_value_data");
    Ok(())
}

#[test]
fn owned_logo_example() {
    let owned_ktx_data: Arc<[u8]> = Arc::from(include_bytes!("babg-bc3.ktx").to_vec());
    let ktx: Ktx<Arc<[u8]>> = Ktx::from(owned_ktx_data);

    assert_eq!(ktx.pixel_width(), 260, "pixel_width");
    assert_eq!(ktx.pixel_height(), 200, "pixel_height");
}

const LOGO_LEVEL_0_BLAKE: &str = "17ae9dcdc7b7f8c38a66fe00ab92759fde35f74cde2aa52449c2ecbca835a51b";
const LOGO_LEVEL_1_BLAKE: &str = "ae05372daa8bb0d45de4431db106e107266307e5a37b51825f10618a9605ee1b";
const LOGO_LEVEL_2_BLAKE: &str = "52ed1d989b8dca91538d68e1077d303a95baf6068d240316bc18c3d8ef5625ea";
const LOGO_LEVEL_3_BLAKE: &str = "3869a1ccc011b1c74255f2a7ccc2a2c5118a492599aea758c59e4b4b80e3bd6a";
const LOGO_LEVEL_4_BLAKE: &str = "5cca1fd6eeb47922490392884e2a5a177f791a37da64109705127b092f4073bc";
const LOGO_LEVEL_5_BLAKE: &str = "2e280b7441e1576d93a7cdf97f67e3fcecdefe209a34c10e9adf5a4158351884";
const LOGO_LEVEL_6_BLAKE: &str = "22703f682061beb020f2316cbcad901268f1ad7869fd1892062b9d50b92508c0";
const LOGO_LEVEL_7_BLAKE: &str = "971ccf807344ecef5e43d10bcd0bd9260d35b1632548d045ba2cbbbf8a50075a";

#[test]
fn include_logo_example_textures() {
    let ktx = include_ktx!("babg-bc3.ktx");
    let mut textures = ktx.textures();

    assert_eq!(
        format!("{:x}", Blake2s::digest(textures.next().unwrap())),
        LOGO_LEVEL_0_BLAKE
    );
    assert_eq!(
        format!("{:x}", Blake2s::digest(textures.next().unwrap())),
        LOGO_LEVEL_1_BLAKE
    );
    assert_eq!(
        format!("{:x}", Blake2s::digest(textures.next().unwrap())),
        LOGO_LEVEL_2_BLAKE
    );
    assert_eq!(
        format!("{:x}", Blake2s::digest(textures.next().unwrap())),
        LOGO_LEVEL_3_BLAKE
    );
    assert_eq!(
        format!("{:x}", Blake2s::digest(textures.next().unwrap())),
        LOGO_LEVEL_4_BLAKE
    );
    assert_eq!(
        format!("{:x}", Blake2s::digest(textures.next().unwrap())),
        LOGO_LEVEL_5_BLAKE
    );
    assert_eq!(
        format!("{:x}", Blake2s::digest(textures.next().unwrap())),
        LOGO_LEVEL_6_BLAKE
    );
    assert_eq!(
        format!("{:x}", Blake2s::digest(textures.next().unwrap())),
        LOGO_LEVEL_7_BLAKE
    );
    assert_eq!(textures.next(), None);
}

#[test]
fn read_logo_example_textures() -> io::Result<()> {
    let ktx = ktx::Decoder::new(BufReader::new(File::open("tests/babg-bc3.ktx")?))?;
    let mut textures = ktx.read_textures();

    assert_eq!(
        format!("{:x}", Blake2s::digest(&textures.next().unwrap())),
        LOGO_LEVEL_0_BLAKE
    );
    assert_eq!(
        format!("{:x}", Blake2s::digest(&textures.next().unwrap())),
        LOGO_LEVEL_1_BLAKE
    );
    assert_eq!(
        format!("{:x}", Blake2s::digest(&textures.next().unwrap())),
        LOGO_LEVEL_2_BLAKE
    );
    assert_eq!(
        format!("{:x}", Blake2s::digest(&textures.next().unwrap())),
        LOGO_LEVEL_3_BLAKE
    );
    assert_eq!(
        format!("{:x}", Blake2s::digest(&textures.next().unwrap())),
        LOGO_LEVEL_4_BLAKE
    );
    assert_eq!(
        format!("{:x}", Blake2s::digest(&textures.next().unwrap())),
        LOGO_LEVEL_5_BLAKE
    );
    assert_eq!(
        format!("{:x}", Blake2s::digest(&textures.next().unwrap())),
        LOGO_LEVEL_6_BLAKE
    );
    assert_eq!(
        format!("{:x}", Blake2s::digest(&textures.next().unwrap())),
        LOGO_LEVEL_7_BLAKE
    );
    assert_eq!(textures.next(), None);
    Ok(())
}

#[test]
fn logo_example_texture_level() {
    let ktx = include_ktx!("babg-bc3.ktx");

    assert_eq!(
        format!("{:x}", Blake2s::digest(ktx.texture_level(0))),
        LOGO_LEVEL_0_BLAKE
    );
    assert_eq!(
        format!("{:x}", Blake2s::digest(ktx.texture_level(4))),
        LOGO_LEVEL_4_BLAKE
    );
}

#[test]
fn logo_example_debug() {
    let dbg_string = format!("{:?}", include_ktx!("babg-bc3.ktx"));
    assert_eq!(
        &dbg_string,
        "Ktx { header: KtxHeader { big_endian: false, gl_type: 0, gl_type_size: 1, gl_format: 0, gl_internal_format: 33779, gl_base_internal_format: 6408, pixel_width: 260, pixel_height: 200, pixel_depth: 0, array_elements: 0, faces: 1, mipmap_levels: 8, bytes_of_key_value_data: 0 } }"
    );
}

#[test]
fn uffizi_6face_include() {
    let ktx = include_ktx!("uffizi_rgba16f_cube.ktx");

    assert!(!ktx.big_endian(), "!big_endian");
    assert_eq!(ktx.gl_type(), 5131, "gl_type");
    assert_eq!(ktx.gl_type_size(), 2, "gl_type_size");
    assert_eq!(ktx.gl_format(), 6408, "gl_format");
    assert_eq!(ktx.gl_internal_format(), 34842, "gl_internal_format");
    assert_eq!(
        ktx.gl_base_internal_format(),
        6408,
        "gl_base_internal_format"
    );
    assert_eq!(ktx.pixel_width(), 512, "pixel_width");
    assert_eq!(ktx.pixel_height(), 512, "pixel_height");
    assert_eq!(ktx.pixel_depth(), 0, "pixel_depth");
    assert_eq!(ktx.array_elements(), 0, "array_elements");
    assert_eq!(ktx.faces(), 6, "faces");
    assert_eq!(ktx.mipmap_levels(), 10, "mipmap_levels");
    assert_eq!(ktx.bytes_of_key_value_data(), 0, "bytes_of_key_value_data");

    assert_eq!(ktx.textures().count(), 10, "ktx.textures().count()");
}

#[test]
fn uffizi_6face_read() {
    let ktx_file = BufReader::new(File::open("tests/uffizi_rgba16f_cube.ktx").unwrap());
    let ktx = ktx::Decoder::new(ktx_file).unwrap();

    assert!(!ktx.big_endian(), "!big_endian");
    assert_eq!(ktx.gl_type(), 5131, "gl_type");
    assert_eq!(ktx.gl_type_size(), 2, "gl_type_size");
    assert_eq!(ktx.gl_format(), 6408, "gl_format");
    assert_eq!(ktx.gl_internal_format(), 34842, "gl_internal_format");
    assert_eq!(
        ktx.gl_base_internal_format(),
        6408,
        "gl_base_internal_format"
    );
    assert_eq!(ktx.pixel_width(), 512, "pixel_width");
    assert_eq!(ktx.pixel_height(), 512, "pixel_height");
    assert_eq!(ktx.pixel_depth(), 0, "pixel_depth");
    assert_eq!(ktx.array_elements(), 0, "array_elements");
    assert_eq!(ktx.faces(), 6, "faces");
    assert_eq!(ktx.mipmap_levels(), 10, "mipmap_levels");
    assert_eq!(ktx.bytes_of_key_value_data(), 0, "bytes_of_key_value_data");

    assert_eq!(
        ktx.read_textures().count(),
        10,
        "ktx.read_textures().count()"
    );
}


src/lib.rs
//! KTX v1 texture storage format parsing.
//!
//! Parses byte data according to
//! [https://www.khronos.org/registry/KTX/specs/1.0/ktxspec_v1.html](https://www.khronos.org/registry/KTX/specs/1.0/ktxspec_v1.html).
//!
//! # Example: Include at compile time
//! ```
//! # fn main() -> std::io::Result<()> {
//! use ktx::{Ktx, include_ktx, KtxInfo};
//!
//! // Include & use static ktx data
//! let image: Ktx<_> = include_ktx!("../tests/babg-bc3.ktx");
//! assert_eq!(image.pixel_width(), 260);
//! # Ok(()) }
//! ```
//!
//! # Example: Read at runtime
//! ```
//! # fn main() -> std::io::Result<()> {
//! # use std::{io::BufReader, fs::File};
//! use ktx::KtxInfo;
//!
//! # let mut buf_reader = BufReader::new(File::open("tests/babg-bc3.ktx").unwrap());
//! let decoder = ktx::Decoder::new(buf_reader)?;
//! assert_eq!(decoder.pixel_width(), 260);
//! # Ok(()) }
//! ```
#![cfg_attr(not(feature = "std"), no_std)]
#![allow(clippy::cast_lossless)]

pub mod header;
#[cfg(feature = "std")]
pub mod read;
pub mod slice;

pub use header::KtxInfo;
#[cfg(feature = "std")]
pub use read::KtxDecoder as Decoder;
pub use slice::Ktx;


src/slice.rs
use crate::header::*;
use byteorder::{BigEndian, ByteOrder, LittleEndian};
use core::{fmt, ops::Deref};

/// KTX texture storage format data stored in a complete slice.
/// Provides [`KtxInfo`](../header/trait.KtxInfo.html).
///
/// See the [specification](https://www.khronos.org/registry/KTX/specs/1.0/ktxspec_v1.html).
///
/// # Example
/// ```
/// # use ktx::*;
/// let image: Ktx<_> = include_ktx!("../tests/babg-bc3.ktx");
/// let texture_levels: Vec<&[u8]> = image.textures().collect();
/// ```
#[derive(Clone, Copy)]
pub struct Ktx<D> {
    header: KtxHeader,
    ktx_data: D,
    texture_start: u32,
}

impl<D> AsRef<KtxHeader> for Ktx<D> {
    #[inline]
    fn as_ref(&self) -> &KtxHeader {
        &self.header
    }
}

impl<D> fmt::Debug for Ktx<D> {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        fmt.debug_struct("Ktx")
            .field("header", &self.header)
            .finish()
    }
}

impl<D> Ktx<D>
where
    D: Deref<Target = [u8]>,
{
    /// Parses a complete KTX data slice and returns a `Ktx` instance.
    #[inline]
    pub fn new(ktx_data: D) -> Self {
        let header = KtxHeader::new(&ktx_data);
        let texture_start = 64 + header.bytes_of_key_value_data();
        Self {
            header,
            ktx_data,
            texture_start,
        }
    }

    /// Returns texture data at the input level, starting at `0`.
    ///
    /// # Panics
    ///
    /// Input level is >= the `mipmap_levels` value.
    #[inline]
    pub fn texture_level(&self, level: u32) -> &[u8] {
        self.textures().nth(level as _).expect("invalid level")
    }

    /// Returns an iterator over the texture levels starting at level 0.
    #[inline]
    pub fn textures(&self) -> Textures<'_, D> {
        Textures {
            parent: self,
            next_level: 0,
            level_end: self.texture_start as _,
        }
    }
}

impl<D> From<D> for Ktx<D>
where
    D: Deref<Target = [u8]>,
{
    #[inline]
    fn from(d: D) -> Self {
        Ktx::new(d)
    }
}

/// Iterator over texture level data.
///
/// For cubemap textures each level will contain all 6 faces
/// in order: +X, -X, +Y, -Y, +Z, -Z.
#[derive(Debug)]
pub struct Textures<'a, D> {
    parent: &'a Ktx<D>,
    next_level: u32,
    level_end: usize,
}

impl<'a, D> Iterator for Textures<'a, D>
where
    D: Deref<Target = [u8]>,
{
    type Item = &'a [u8];

    fn next(&mut self) -> Option<Self::Item> {
        if self.next_level >= self.parent.mipmap_levels() {
            None
        } else {
            self.next_level += 1;

            let l_end = self.level_end;
            let mut next_lvl_len = if self.parent.big_endian() {
                BigEndian::read_u32(&self.parent.ktx_data[l_end..l_end + 4])
            } else {
                LittleEndian::read_u32(&self.parent.ktx_data[l_end..l_end + 4])
            };

            if self.parent.array_elements() == 0 && self.parent.faces() == 6 {
                // Multiply for each face, see https://www.khronos.org/registry/KTX/specs/1.0/ktxspec_v1.html#2.16
                next_lvl_len *= 6;
            }

            self.level_end = l_end + 4 + next_lvl_len as usize;
            Some(&self.parent.ktx_data[l_end + 4..self.level_end])
        }
    }
}

impl<D> core::iter::FusedIterator for Textures<'_, D> where D: Deref<Target = [u8]> {}

/// Wrapper for `include_bytes!` returning `Ktx<'static [u8]>`
///
/// # Example
/// ```
/// use ktx::{include_ktx, Ktx};
/// let image: Ktx<&'static [u8]> = include_ktx!("../tests/babg-bc3.ktx");
/// ```
#[macro_export]
macro_rules! include_ktx {
    ($path:tt) => {
        $crate::Ktx::new(include_bytes!($path) as &[u8])
    };
}


src/write.rs
use crate::header::{KtxHeader, KtxInfo, KTX_IDENTIFIER};
use byteorder::{BigEndian, ByteOrder, LittleEndian};

pub struct KtxBuilder {
    header: KtxHeader,
    levels: Vec<Vec<u8>>,
}
impl KtxBuilder {
    pub fn new(header: KtxHeader) -> Self {
        Self {
            header,
            levels: Vec::default(),
        }
    }

    pub fn with_level(mut self, texture: Vec<u8>) -> Self {
        self.add_level(texture);

        self
    }

    pub fn add_level(&mut self, texture: Vec<u8>) {
        self.levels.push(texture);
    }

    pub fn to_vec(self) -> Result<Vec<u8>, &'static str> {
        let KtxBuilder { header, levels } = self;

        // allocate the full size
        let size = KTX_IDENTIFIER.len()
            + std::mem::size_of::<KtxHeader>()
            + levels
                .iter()
                .map(|level| level.len() + std::mem::size_of::<u32>())
                .sum::<usize>();

        let mut buffer = Vec::with_capacity(size);
        buffer.resize(size, 0);

        header.write(&mut buffer[0..64]);

        let mut cur_index = 64;

        for level in levels {
            let write_end = level.len() as u32 / header.faces();
            let cur_end = cur_index + level.len() + std::mem::size_of::<u32>();

            if header.big_endian() {
                BigEndian::write_u32_into(
                    &[write_end],
                    &mut buffer[cur_index..cur_index + std::mem::size_of::<u32>()],
                );
            } else {
                LittleEndian::write_u32_into(
                    &[write_end],
                    &mut buffer[cur_index..cur_index + std::mem::size_of::<u32>()],
                )
            }

            buffer[cur_index + std::mem::size_of::<u32>()..cur_end].copy_from_slice(&level);

            cur_index = cur_end;
        }

        Ok(buffer)
    }
}


src/header.rs
use byteorder::{BigEndian, ByteOrder, LittleEndian};

pub(crate) const KTX1_IDENTIFIER: [u8; 12] = [
    0xAB, 0x4B, 0x54, 0x58, 0x20, 0x31, 0x31, 0xBB, 0x0D, 0x0A, 0x1A, 0x0A,
];

/// KTX texture storage format parameters.
///
/// See the [specification](https://www.khronos.org/registry/KTX/specs/1.0/ktxspec_v1.html).
pub trait KtxInfo {
    /// endianness contains the number 0x04030201 written as a 32 bit integer. If the file is little
    /// endian then this is represented as the bytes 0x01 0x02 0x03 0x04. If the file is big endian
    /// then this is represented as the bytes 0x04 0x03 0x02 0x01. When reading endianness as a 32
    /// bit integer produces the value 0x04030201 then the endianness of the file matches the the
    /// endianness of the program that is reading the file and no conversion is necessary. When
    /// reading endianness as a 32 bit integer produces the value 0x01020304 then the endianness of
    /// the file is opposite the endianness of the program that is reading the file, and in that
    /// case the program reading the file must endian convert all header bytes and, if glTypeSize >
    /// 1, all texture data to the endianness of the program (i.e. a little endian program must
    /// convert from big endian, and a big endian program must convert to little endian).
    fn big_endian(&self) -> bool;
    /// For compressed textures, glType must equal 0. For uncompressed textures, glType specifies the
    /// type parameter passed to glTex{,Sub}Image*D, usually one of the values from table 8.2 of the
    /// OpenGL 4.4 specification [OPENGL44] (UNSIGNED_BYTE, UNSIGNED_SHORT_5_6_5, etc.)
    ///
    /// [OPENGL44]: https://www.khronos.org/registry/KTX/specs/1.0/ktxspec_v1.html#refsGL44
    fn gl_type(&self) -> u32;
    /// glTypeSize specifies the data type size that should be used when endianness conversion is
    /// required for the texture data stored in the file. If glType is not 0, this should be the
    /// size in bytes corresponding to glType. For texture data which does not depend on platform
    /// endianness, including compressed texture data, glTypeSize must equal 1.
    fn gl_type_size(&self) -> u32;
    /// For compressed textures, glFormat must equal 0. For uncompressed textures, glFormat specifies
    /// the format parameter passed to glTex{,Sub}Image*D, usually one of the values from table 8.3
    /// of the OpenGL 4.4 specification [OPENGL44] (RGB, RGBA, BGRA, etc.)
    ///
    /// [OPENGL44]: https://www.khronos.org/registry/KTX/specs/1.0/ktxspec_v1.html#refsGL44
    fn gl_format(&self) -> u32;
    /// For compressed textures, glInternalFormat must equal the compressed internal format, usually
    /// one of the values from table 8.14 of the OpenGL 4.4 specification [OPENGL44]. For
    /// uncompressed textures, glInternalFormat specifies the internalformat parameter passed to
    /// glTexStorage*D or glTexImage*D, usually one of the sized internal formats from tables 8.12 &
    /// 8.13 of the OpenGL 4.4 specification [OPENGL44]. The sized format should be chosen to match
    /// the bit depth of the data provided. glInternalFormat is used when loading both compressed
    /// and uncompressed textures, except when loading into a context that does not support sized
    /// formats, such as an unextended OpenGL ES 2.0 context where the internalformat parameter is
    /// required to have the same value as the format parameter.
    ///
    /// [OPENGL44]: https://www.khronos.org/registry/KTX/specs/1.0/ktxspec_v1.html#refsGL44
    fn gl_internal_format(&self) -> u32;
    /// For both compressed and uncompressed textures, glBaseInternalFormat specifies the base
    /// internal format of the texture, usually one of the values from table 8.11 of the OpenGL 4.4
    /// specification [OPENGL44] (RGB, RGBA, ALPHA, etc.). For uncompressed textures, this value
    /// will be the same as glFormat and is used as the internalformat parameter when loading into a
    /// context that does not support sized formats, such as an unextended OpenGL ES 2.0 context.
    ///
    /// [OPENGL44]: https://www.khronos.org/registry/KTX/specs/1.0/ktxspec_v1.html#refsGL44
    fn gl_base_internal_format(&self) -> u32;
    /// The width of the texture image for level 0, in pixels. No rounding to block sizes should be
    /// applied for block compressed textures.
    fn pixel_width(&self) -> u32;
    /// The height of the texture image for level 0, in pixels. No rounding to block sizes should be
    /// applied for block compressed textures.
    ///
    /// For 1D textures this must be 0.
    fn pixel_height(&self) -> u32;
    /// The depth of the texture image for level 0, in pixels. No rounding to block sizes should be
    /// applied for block compressed textures.
    ///
    /// For 1D textures this must be 0. For 2D and cube textures this must be 0.
    fn pixel_depth(&self) -> u32;
    /// numberOfArrayElements specifies the number of array elements. If the texture is not an array
    /// texture, numberOfArrayElements must equal 0.
    fn array_elements(&self) -> u32;
    /// numberOfFaces specifies the number of cubemap faces. For cubemaps and cubemap arrays this
    /// should be 6. For non cubemaps this should be 1. Cube map faces are stored in the order: +X,
    /// -X, +Y, -Y, +Z, -Z.
    ///
    /// Due to GL_OES_compressed_paletted_texture [OESCPT] not defining the interaction between
    /// cubemaps and its GL_PALETTE* formats, if `glInternalFormat` is one of its GL_PALETTE*
    /// format, numberOfFaces must be 1
    ///
    /// [OESCPT]: https://www.khronos.org/registry/KTX/specs/1.0/ktxspec_v1.html#refsOESCPT
    fn faces(&self) -> u32;
    /// numberOfMipmapLevels must equal 1 for non-mipmapped textures. For mipmapped textures, it
    /// equals the number of mipmaps. Mipmaps are stored in order from largest size to smallest
    /// size. The first mipmap level is always level 0. A KTX file does not need to contain a
    /// complete mipmap pyramid. If numberOfMipmapLevels equals 0, it indicates that a full mipmap
    /// pyramid should be generated from level 0 at load time (this is usually not allowed for
    /// compressed formats).
    fn mipmap_levels(&self) -> u32;
    /// keyAndValueByteSize is the number of bytes of combined key and value data in one key/value
    /// pair following the header. This includes the size of the key, the NUL byte terminating the
    /// key, and all the bytes of data in the value. If the value is a UTF-8 string it should be NUL
    /// terminated and the keyAndValueByteSize should include the NUL character (but code that reads
    /// KTX files must not assume that value fields are NUL terminated). keyAndValueByteSize does
    /// not include the bytes in valuePadding.
    fn bytes_of_key_value_data(&self) -> u32;
}

/// KTX texture storage format header. Provides [`KtxInfo`](../header/trait.KtxInfo.html).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct KtxHeader {
    big_endian: bool,
    gl_type: u32,
    gl_type_size: u32,
    gl_format: u32,
    gl_internal_format: u32,
    gl_base_internal_format: u32,
    pixel_width: u32,
    pixel_height: u32,
    pixel_depth: u32,
    array_elements: u32,
    faces: u32,
    mipmap_levels: u32,
    bytes_of_key_value_data: u32,
}

impl KtxHeader {
    /// Reads first 64 bytes to parse KTX header data, returns a `KtxHeader`.
    pub fn new(first_64_bytes: &[u8]) -> Self {
        debug_assert!(first_64_bytes.len() >= 64);
        debug_assert_eq!(&first_64_bytes[..12], &KTX1_IDENTIFIER, "Not KTX1");

        let big_endian = first_64_bytes[12] == 4;

        let mut vals: [u32; 12] = <_>::default();
        if big_endian {
            BigEndian::read_u32_into(&first_64_bytes[16..64], &mut vals);
        } else {
            LittleEndian::read_u32_into(&first_64_bytes[16..64], &mut vals);
        }

        Self {
            big_endian,
            gl_type: vals[0],
            gl_type_size: vals[1],
            gl_format: vals[2],
            gl_internal_format: vals[3],
            gl_base_internal_format: vals[4],
            pixel_width: vals[5],
            pixel_height: vals[6],
            pixel_depth: vals[7],
            array_elements: vals[8],
            faces: vals[9],
            mipmap_levels: vals[10],
            bytes_of_key_value_data: vals[11],
        }
    }
}

impl AsRef<KtxHeader> for KtxHeader {
    #[inline]
    fn as_ref(&self) -> &Self {
        self
    }
}

impl<T> KtxInfo for T
where
    T: AsRef<KtxHeader>,
{
    #[inline]
    fn big_endian(&self) -> bool {
        self.as_ref().big_endian
    }
    #[inline]
    fn gl_type(&self) -> u32 {
        self.as_ref().gl_type
    }
    #[inline]
    fn gl_type_size(&self) -> u32 {
        self.as_ref().gl_type_size
    }
    #[inline]
    fn gl_format(&self) -> u32 {
        self.as_ref().gl_format
    }
    #[inline]
    fn gl_internal_format(&self) -> u32 {
        self.as_ref().gl_internal_format
    }
    #[inline]
    fn gl_base_internal_format(&self) -> u32 {
        self.as_ref().gl_base_internal_format
    }
    #[inline]
    fn pixel_width(&self) -> u32 {
        self.as_ref().pixel_width
    }
    #[inline]
    fn pixel_height(&self) -> u32 {
        self.as_ref().pixel_height
    }
    #[inline]
    fn pixel_depth(&self) -> u32 {
        self.as_ref().pixel_depth
    }
    #[inline]
    fn array_elements(&self) -> u32 {
        self.as_ref().array_elements
    }
    #[inline]
    fn faces(&self) -> u32 {
        self.as_ref().faces
    }
    #[inline]
    fn mipmap_levels(&self) -> u32 {
        self.as_ref().mipmap_levels
    }
    #[inline]
    fn bytes_of_key_value_data(&self) -> u32 {
        self.as_ref().bytes_of_key_value_data
    }
}


src/read.rs
use crate::header::*;
use byteorder::{BigEndian, ByteOrder, LittleEndian};
use std::{
    fmt,
    io::{self, Read},
};

/// KTX texture storage format reader. Useful when reading from a file and/or compressed data.
/// Provides [`KtxInfo`](../header/trait.KtxInfo.html).
///
/// # Example
/// ```
/// # use std::{io::BufReader, fs::File};
/// # fn main() -> std::io::Result<()> {
/// use ktx::*;
/// # let mut buf_reader = BufReader::new(File::open("tests/babg-bc3.ktx")?);
/// let mut decoder = ktx::Decoder::new(buf_reader)?;
///
/// assert_eq!(decoder.pixel_width(), 260);
/// let texture_levels: Vec<Vec<u8>> = decoder.read_textures().collect();
/// # Ok(()) }
/// ```
#[derive(Clone, Copy)]
pub struct KtxDecoder<R> {
    header: KtxHeader,
    data: R,
}

impl<R> AsRef<KtxHeader> for KtxDecoder<R> {
    #[inline]
    fn as_ref(&self) -> &KtxHeader {
        &self.header
    }
}

impl<R> fmt::Debug for KtxDecoder<R> {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        fmt.debug_struct("KtxDecoder")
            .field("header", &self.header)
            .finish()
    }
}

impl<R: io::Read> KtxDecoder<R> {
    /// Reads KTX header data and returns a `KtxDecoder`.
    #[inline]
    pub fn new(mut data: R) -> io::Result<Self> {
        let mut header_data = [0; 64];
        data.read_exact(&mut header_data)?;
        let header = KtxHeader::new(&header_data);
        Ok(Self { header, data })
    }

    /// Consumes the `KtxDecoder` to returns an iterator reading texture levels starting at level 0.
    #[inline]
    pub fn read_textures(self) -> Textures<R> {
        Textures {
            header: self.header,
            data: self.data,
            next_level: 0,
        }
    }

    /// Returns `KtxHeader`. Useful if this info is desired after consuming the `KtxDecoder`.
    ///
    /// # Example
    /// ```
    /// # use std::{io::BufReader, fs::File};
    /// # use ktx::*;
    /// # let mut reader = BufReader::new(File::open("tests/babg-bc3.ktx").unwrap());
    /// # let mut decoder = ktx::Decoder::new(reader).unwrap();
    /// let ktx_header = decoder.header();
    /// assert_eq!(ktx_header.pixel_width(), 260);
    /// ```
    #[inline]
    pub fn header(&self) -> KtxHeader {
        self.header
    }
}

/// Iterator that reads texture level data into `Vec<u8>`.
///
/// For cubemap textures each level will contain all 6 faces
/// in order: +X, -X, +Y, -Y, +Z, -Z.
#[derive(Debug)]
pub struct Textures<R> {
    header: KtxHeader,
    data: R,
    next_level: u32,
}

impl<R: io::Read> Iterator for Textures<R> {
    type Item = Vec<u8>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.next_level >= self.header.mipmap_levels() {
            None
        } else {
            // skip key-value data
            if self.next_level == 0 && self.header.bytes_of_key_value_data() != 0 {
                let mut discard = Vec::with_capacity(self.header.bytes_of_key_value_data() as _);
                self.data
                    .by_ref()
                    .take(self.header.bytes_of_key_value_data() as _)
                    .read_to_end(&mut discard)
                    .ok()?;
            }

            self.next_level += 1;
            let mut level_len = {
                let mut len = [0; 4];
                self.data.read_exact(&mut len).ok()?;
                if self.header.big_endian() {
                    BigEndian::read_u32(&len)
                } else {
                    LittleEndian::read_u32(&len)
                }
            };

            if self.header.array_elements() == 0 && self.header.faces() == 6 {
                // Multiply for each face, see https://www.khronos.org/registry/KTX/specs/1.0/ktxspec_v1.html#2.16
                level_len *= 6;
            }

            let mut level = Vec::with_capacity(level_len as _);
            self.data
                .by_ref()
                .take(level_len as _)
                .read_to_end(&mut level)
                .ok()?;
            Some(level)
        }
    }
}

impl<R: io::Read> std::iter::FusedIterator for Textures<R> {}
